### DISCUSSION TITLE
Trouble with Static Files and Components in Django Deployment

### DISCUSSION THREAD

**Discussion by dnl0037** [reactive-python/reactpy-django#166] [Problem] [answered]
Hello! After a bit of research and work, I hooked up my ReactPy components to Django, and I'm running deployment tests. However, I've run into the inconvenience that the static files (such as the CSS style) have a peculiar way of being handled, which I am not doing yet, and thus they don't load. But the ReactPy components won't load either. Should I treat them as static files?

**Comment by Archmonger** [comment]
Are you using [ReactPy-Django](https://github.com/reactive-python/reactpy-django)? If so, you can use the [Django CSS component](https://reactive-python.github.io/reactpy-django/features/components/#django-css).

**Reply by Archmonger** [reply]
If this is not your issue, I'm going to need more details about what specific problems you're having.
For example you said sometimes your components don't load, are you getting any errors in console?

**Comment by dnl0037** [comment]
Hi. Yes, I am using ReactPy-Django. In my computer, everything loads fine, but now I uploaded my project to "pythonanywhere.com". Here, the login (which is not a component) loads but the main page (which is a component) does not. It throws me this error that I think is causing the problem:
Not Found: /static/reactpy_django/client.js

**Reply by Archmonger** [reply]
Did you make sure to run `python manage.py collectstatic`?

**Reply by Archmonger** [reply]
Additionally, when running in production (`settings.py:DEBUG = False`) Django has no built-in way of serving static files.

Most Django developers end up using [Whitenoise](https://whitenoise.readthedocs.io/en/latest/) to fill this gap.

Another common configuration involves serving static files using Nginx.

**Comment by dnl0037** [comment]
Yep, I run the "collectstatic" command and started using the "whitenoise" library. I placed the components in the static file alongside the stylesheets. Now the CSS works, including the CSS for the components (I can see the background color), but the components themselves don't load. I'm getting these kinds of errors:
2023-08-04 22:52:06,437: Not Found: /reactpy/static.components.adminpage/1fc84cc50c1d422e9482a96a9c3be94f/

My code for loading the component is:
```
{% block content %}
    {% component "staticfiles.components.adminpage" user %}
{% endblock %}
```

**Reply by Archmonger** [reply]
Your python files should not be contained within a static files directory.
Please put them directly within your Django project.

**Reply by dnl0037** [reply]
Hello again. I think I now know how to handle the static part, but unfortunately, I cannot get my components to load. It might be related to the REACTPY_WEBSOCKET_URL="reactpy/" because I keep encountering these errors:

`Not Found: /reactpy/myapp.components.hello_world/daedb19f6dca46aebb9e4c6feebe19c5/`

And when I inspect the webpage where the component should be and go to script, I see this:

 ```
 import { mountViewToElement } from "/static/reactpy_django/client.a6480d8362bf.js";
  const mountElement = document.getElementById("daedb19f6dca46aebb9e4c6feebe19c5");
  mountViewToElement(
    mountElement,
    "reactpy/",
    "reactpy/web_module/",
    "259200",
    "myapp.components.hello_world/daedb19f6dca46aebb9e4c6feebe19c5/",
  );
```
The ID seems to be similar, so maybe there is some sort of connection.

I have never deployed an app before and probably I am missing a lot of things or messing up my configuration. By any chance do you have a preferred server provider  or any deployed Reactpy-Django app configuration examples that I can refer to?
I apologize for asking you so many questions, I really appreciate your time and help.

**Reply by Archmonger** [reply]
Can you install the latest version (3.3.0) of `reactpy-django`? I've added a ton of automated checks and they might catch some of your configuration issues.

These checks only run if `settings.py:DEBUG = True`.

**Reply by Archmonger** [reply]
If that doesn't catch all your issues, I might need to see your repo if possible.

**Reply by dnl0037** [reply]
I would really appreciate if you take a look: https://github.com/dnl0037/djangorp. The component.py is inside the main directory.

**Reply by Archmonger** [reply, marked as answer]
Seems to work for me with `DEBUG` off. If you're having issues with remote hosting, you might need to look into WebSocket configuration with your hosting provider.

**Reply by dnl0037** [reply]
Thank you very much for checking it out. I just searched for the WebSockets configuration of the "PythonAnywhere" server provider, and they said that they do not support them... I learned the hard way, haha. I will now look for another provider. Thank you again for your help!

**Reply by dnl0037** [reply]
Hey, I wanted to let you know that when I installed the lastest version of reactpy-django, the app started to be a bit laggy. With all the changes I was making I did't know why it was happening. But a few minutes ago I returned to the previous version and it began to run smoothly again.

**Reply by Archmonger** [reply]
If the issue is first-render time, ReactPy now uses the database to keep track of when it needs to perform cleaning. Therefore, a slow database speed will result in slow first render times. You can try a higher performance database to fix this.

If the issue is re-render time (such as click events), you can try turning off `settings.py:REACTPY_BACKHAUL_THREAD`.

**Reply by dnl0037** [reply]
I see. I've been using the database that django provided, sqlite3 (and only for saving the users, which are maybe 2 or 3.) If I set "REACTPY_DATABASE = "default", does Reactpy connect itself to it?

**Reply by Archmonger** [reply]
Yep. However, you don't need to set `REACTPY_DATABASE = "default"` since that's the default behavior.

**Reply by Archmonger** [reply]
Can you try `REACTPY_BACKHAUL_THREAD = False` to see if that fixes your performance issues?

**Reply by dnl0037** [reply]
Hi, I hadn't seen the latest messages. Yes, with REACTPY_BACKHAUL_THREAD = False, the code is running smoothly again. Thank you

**Reply by Archmonger** [reply]
@dnl0037 I was not able to replicate the performance issues you saw with `REACTPY_BACKHAUL_THREAD`. Can you test out this new build for me and see if it fixes the performance?

`pip install git+https://github.com/Archmonger/reactpy-django@3.3.2`

**Reply by dnl0037** [reply]
Hi. Yeah, I installed the package and the code runs smoothly whether I add the REACTPY_BACKHAUL_THREAD = False or not.

**Reply by dnl0037** [reply]
should I keep that version?

**Reply by Archmonger** [reply]
Can you try explicitly adding `REACTPY_BACKHAUL_THREAD = True`?

**Reply by dnl0037** [reply]
With REACTPY_BACKHAUL_THREAD = True, it becomes slightly sluggish again.

**Reply by Archmonger** [reply]
Okay, can you keep `REACTPY_BACKHAUL_THREAD = True`, but then uninstall & reinstall `reactpy_django`? This issue might be related to `__pycache__` optimizations.

1. `pip uninstall reactpy_django`
2. `pip install reactpy_django==3.3.1`

**Reply by dnl0037** [reply]
Still laggy :/ 
Maybe it's something about the configuration? I have this:
REACTPY_CACHE = "default"
REACTPY_RECONNECT_MAX = 259200
REACTPY_WEBSOCKET_URL = "reactpy/"
REACTPY_DEFAULT_QUERY_POSTPROCESSOR = "reactpy_django.utils.django_query_postprocessor"
REACTPY_BACKHAUL_THREAD = True

**Reply by Archmonger** [reply]
Those settings look good.

I just tried to create a different workaround for `REACTPY_BACKHAUL_THREAD`, can you try this one out?

1. `pip uninstall reactpy_django`
2. `pip install git+https://github.com/Archmonger/reactpy-django@3.3.2`

Also, please let me know what Python version and operating system you are using.

**Reply by dnl0037** [reply]
Im afraid it's still laggy. I use windows 11 and python 3.11

**Reply by Archmonger** [reply]
Are you using a Python venv? If so can you recreate your venv?

**Reply by dnl0037** [reply]
Hi! I'm sorry I missed the last message. Yes, I'm using a venv, but I'm not sure what you mean by recreating it

**Reply by Archmonger** [reply]
To recreate it you need to delete your venv folder and then remake it.

**Reply by dnl0037** [reply]
Got it, done. But if REACTPY_BACKHAUL_THREAD = True, it keeps being laggy

**Reply by Archmonger** [reply]
@dnl0037 How are you running Django when you're checking if it's laggy?

Have you checked if it's laggy when running Django through a webserver like `hypercorn` or `uvicorn`?

**Reply by dnl0037** [reply]
Actually, I'm not sure. I haven't deployed yet, and I'm still running the code locally with 'python manage.py runserver'. 
Hey, by the way, this isn't a big deal for me, I don't want to make you lose time on something that might not be in the interest of the library.

**Reply by Archmonger** [reply]
The backhaul implementation is eventually going to get merged into core, so it's useful to find out how to replicate these performance problems early on.

Instead of using `runserver` in your local environment, can you try
1. Enable backhaul threads
2. `pip install hypercorn`
3. `hypercorn example.path.to.asgi:application`

If this doesn't work I'll give up on solving this for now.

**Reply by dnl0037** [reply]
I understand. Ok, good news, I've installed Hypercorn and flipped the switch for 'REACTPY_BACKHAUL_THREAD = True'. After starting the 'hypercorn app:application' command, everything seems to be running pretty smoothly.
